### ISSUE TITLE
pMHCI data exploration - BA quantitative and = only

### ISSUE THREAD

**Issue by gcroci2** [DeepRank/3D-Vac#113]
It's useful to have a detailed overview of the new data we have before starting training the models. The data main metafeatures' are described in `/projects/0/einf2380/data/external/processed/I/BA_pMHCI_human_quantitative_only_eq.csv` (updated 29/03/2023) and their 3D models generated with PANDORA are in `/projects/0/einf2380/data/pMHCI/features_input_folder/HLA_quantitative`, together with their pssms. The notebook that I used to explore the data is `src/3_build_db4/GNN/0_metafeatures_exploration.ipynb`.

### 1. Missing data points

<img width="583" alt="image" src="https://user-images.githubusercontent.com/55382553/228539010-6f1a90ee-2903-440c-8ba5-42b0ea2ec0a3.png">

- [ ] Why are they missing?

From now on the information refers to the actual data in the hdf5 files, i.e. only the data for which we do have the PDBs models (100178 total data points).

### 2. Measurement type, kind and inequalities

We are using only data with `quantitative` measurement type, and `affinity` measurement kind, with measurement_inequality only equal to `=`.

In the bigger dataset, we have different measurement inequalities (`>`, `<`, `=`). For now we decided to include only the ones above mentioned.  The inequalities mean that that measure is either > or < than the reported value. This can make the data very noisy, since such values are not well determined.

### 3. Allele types and peptides' lengths

<img width="112" alt="image" src="https://user-images.githubusercontent.com/55382553/213143340-66d7cd62-1a36-4594-b56e-7cc10567d2c1.png">

In the CSV, the peptides' lengths span from 8 to 21, even if we should have peptides with a maximum length of 15. I checked the original CSV file taken from MHCFlurry 2.0 (S3, you can find it [here](https://data.mendeley.com/datasets/zx3kjzc3yx/3)) and there are no peptides longer than 15 there. The peptides with length > 15 are 88 data points:
<img width="539" alt="image" src="https://user-images.githubusercontent.com/55382553/213143514-76b7f178-1d36-448f-b228-969724234ca5.png">

Heleen used a different csv to filter those data (from MHCFlurry GitHub repo) and probably there is no filtering for peptides > 15. Anyway, we decided to not filter them out, because they are not a significant number of data points so they should have basically no influence on the training. Addionally, since we are trying to develop a model which should generalize according to the complexes' structure, it may be actually useful to not filter them out thus increasing peptides' length variability (if they have an influence at all).

### 4. Peptides' clusters

The file `/projects/0/einf2380/data/external/processed/I/BA_pMHCI_human_quantitative_all_hla_gibbs_clusters.csv` contains the peptides clusters' sets created using [GibbsCluster](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5570237/). 

`cluster_set_10` (containing clusters 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10) has been chosen for the peptides clustering experiments. The cluster name which resulted in the most distant one is the one with the name 3.
- [ ] Which is the garbage cluster?

The clusters not included in the test (including NaN ones) should be shuffled between validation and training sets.

If we get bad performances this way, or maybe at a later stage we can a priori try this, we can clusterize the validation and the training sets as well. We'll have a more indicative idea during the training phase of how well the network is generalizing, and it may help the network itself. On the other hand, if we have few data for the test cluster, it may be tricky to do such thing and have enough data for testing (that's the main reason why usually training and validation set are shuffled and not clusterized as well).

#### General data overview

<img width="2522" alt="image" src="https://user-images.githubusercontent.com/55382553/228581121-9ebb8525-b5a4-46c7-8ccd-592c2ea6a05d.png">

#### Clustering on peptides

<img width="2517" alt="image" src="https://user-images.githubusercontent.com/55382553/228587443-9e14a519-a2e4-4498-84f5-8681b89adc29.png">

Using cluster value 3 for the testing set (from `cluster_set_10`, as mentioned above).
Training set: 92556 samples, 92%
	- Class 0: 51756 samples, 56%
	- Class 1: 40800 samples, 44%
Testing set: 7622 samples, 8%
	- Class 0: 4320 samples, 57%
	- Class 1: 3302 samples, 43%

#### Clustering on alleles
- [ ] More information are needed about how the clustering was performed. 

<img width="2517" alt="image" src="https://user-images.githubusercontent.com/55382553/228587595-e48ca962-f353-4ed9-bef0-802a87bea77d.png">

Using cluster value 1 for the testing set.
Training set: 89779 samples, 90%
	- Class 0: 49958 samples, 56%
	- Class 1: 39821 samples, 44%
Testing set: 10399 samples, 10%
	- Class 0: 6118 samples, 59%
	- Class 1: 4281 samples, 41%

#### Experiments

I'll refer here to the **validation set** as the one used for evaluation _during_ training and to the **testing set** as the one used _after_ training.

- We decided that during experiments we're going to start with no cross-validation, from the worst case scenario possible (the test set must be the most distant one from the training(+validation) set). 

Possible general phases for the first rounds:
1.1: data standardization
1.2: weighted the loss function (only for classification)
1.3: increase the net size
1.4: improve features standardization
1.5 same flow with regression task. Note that if we will need MS data as well, since they are only 0s and 1s, we may need to keep classification for adding those data. If results are already good with BA data only, we may want to switch to regression to match the state of the art. Discussion on this point is welcomed. 
1.6: cross-validate the best one (at least 2-folds, we need to decide which cluster will be placed in the second fold - the second most distant one) experiment

- Experiment 1: data shuffling stratifying on target.
- Experiment 2: data will be divided according to clusters, leaving the most distant ones in the test set. 
- Experiment 3: data will be divided according to alleles, leaving the rare ones in the test set (see the image above for more details). 

After this rounds of experiments we can evaluate and investigate the features importance. 

In general, most of the experiments should be done on the shuffled data, and when we find out the "best" general configuration we can try out the clustering experiments (2 and 3). This is of course not the only way of proceeding (which depends a lot on what you want to prove), but so far we decided to move in this direction.
 
We discussed such experiment at the last meetings, but please let me know what do you thing about this planning :)
